docs/src/content/docs/docs/rfcs/0001-direct-storage-architecture.mdx

```diff
@@ -0,0 +1,390 @@
+
+  // The state of this SyncOp.
+  SyncState state = 4;
+
+  // The full remote commit we are attempting to sync.
+  // Only set when `state == Pending`.
+  optional graft.remote.v1.Commit commit = 5;
+}
+
+// stored at log/{vid}/{lsn}
+message LocalCommit {
+  // The Volume ID and LSN.
+  graft.remote.v1.VolumeRef ref = 1;
+
+  // The Volume's page count after replaying this commit.
+  uint32 page_count = 2;
+}
+```
+
+### Algorithms
+
+**TODO: open thoughts**
+- should sync status include an error message? or is code enough?
+- do we want a log of failed syncs stored separately from the sync log?
+  - an app basically only cares how far behind we are, and whether or not the most recent sync failed (and why)
+- do we need to track the min lsn on the writer vid to ensure monotonicity?
+
+**TODO: plan for sync protocol**
+- snapshot
+  - use last writer lsn if exists, otherwise upstream lsn
+- push
+  - build and push segments
+  - write pending SyncOp
+  - commit
+  - if success:
+    - fjall batch
+      - write commit to parent log
+      - mark SyncOp as success
+  - if transient failure:
+    - mark SyncOp as failure, schedule retry
+  - if reject:
+    - mark SyncOp as reject, disable push, wait for app resolution
+- pull
+  - pull commits into remote vid
+  - also write commit into writer vid if exists
+    - if so, add a SyncOp mapping the remote lsn to the writer lsn
+- fast forward
+  - when read snapshots close, check to see if we can fast forward the writer
+  - fjall batch
+    - drop local commits up to the oldest read snapshot
+    - optimization: before dropping segments, best-effort copy cached pages into the upstream VIDs segments by following the mapping in the sync log.
+      - unless we do this during push
+    - drop any unreachable pages
+    - drop SyncOps up to new base LSN
+    - update writer vid parent ref
+- reset
+  - open new writer vid
+  - update writer pointer on upstream vid
+  - eventually GC old vid once read snapshots drop
+
+## Drawbacks
+
+Why should we *not* do this?
+
+## Rationale and alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+## Prior art
+
+Discuss prior art, both the good and the bad, in relation to this proposal. A few examples of what this can include are:
+
+- Does this feature exist in other projects and what experience has their community had?
+- Papers: Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.
+
+This section is intended to encourage you as an author to think about the lessons from other projects, provide readers of your RFC with a fuller picture. If there is no prior art, that is fine - your ideas are interesting to us whether they are brand new or inspired.
```

Comment:
In https://www.localfirst.fm/18/transcript, the section starting at "0:34:25: Why using HTTP for network layer?" was really interesting to me. There is a suggestion that if you avoid using websockets, and use plain HTTP with long-polling instead, you can put a CDN in front of your sync backend, and it will do request coalescing for you.

Reply:
This is a solid point and part of the reason I went with http/protobuf + polling rather than websockets. It should be straightforward to accelerate everything via CDN caches without too much extra work.

Will probably also experiment with long-polling vs websockets for the signaling layer. From reading the [configurator paper](https://research.facebook.com/publications/holistic-configuration-management-at-facebook/) - they found that push based signaling worked better for servers but didn't scale well out to devices. So they have a hybrid model that I think makes  a lot of sense at high scale.